libs/api/libbpf_program.cpp

```diff
@@ -0,0 +1,236 @@
+
+    if (program_type == nullptr) {
+        return EBPF_INVALID_ARGUMENT;
+    }
+
+    const char* log_buffer;
+    return (int)ebpf_program_load(
+        file_name, program_type, nullptr, EBPF_EXECUTION_ANY, object, (fd_t*)program_fd, &log_buffer);
+}
+
+int
+bpf_program__fd(const struct bpf_program* program)
+{
+    return (int)ebpf_program_get_fd(program);
+}
+
+const char*
+bpf_program__name(const struct bpf_program* program)
+{
+    return program->program_name;
+}
+
+const char*
+bpf_program__section_name(const struct bpf_program* program)
+{
+    return program->section_name;
+}
+
+size_t
+bpf_program__size(const struct bpf_program* program)
+{
+    return program->byte_code_size;
+}
+
+struct bpf_link*
+bpf_program__attach(struct bpf_program* program)
+{
+    if (program == nullptr) {
+        return nullptr;
+    }
+
+    ebpf_handle_t link_handle;
+    uint32_t result = ebpf_api_link_program(program->handle, program->attach_type, &link_handle);
+    if (result) {
+        errno = result;
+        return nullptr;
+    }
+
+    // The bpf_link structure is opaque so we can use the link handle directly.
+    return (struct bpf_link*)link_handle;
+}
+
+struct bpf_link*
+bpf_program__attach_xdp(struct bpf_program* program, int ifindex)
+{
+    if (program == nullptr) {
+        return nullptr;
+    }
+
+    // TODO: use ifindex
+    UNREFERENCED_PARAMETER(ifindex);
+
+    ebpf_handle_t link_handle;
+    uint32_t result = ebpf_api_link_program(program->handle, EBPF_ATTACH_TYPE_XDP, &link_handle);
+    if (result) {
+        return nullptr;
+    }
+
+    // The bpf_link structure is opaque so we can use the link handle directly.
+    return (struct bpf_link*)link_handle;
+}
+
+struct bpf_program*
+bpf_program__next(struct bpf_program* previous, const struct bpf_object* object)
+{
+    return ebpf_program_next(previous, object);
+}
+
+struct bpf_program*
+bpf_program__prev(struct bpf_program* next, const struct bpf_object* object)
+{
+    return ebpf_program_previous(next, object);
+}
+
+int
+bpf_program__unpin(struct bpf_program* program, const char* path)
+{
+    char fullpath[MAX_PATH];
+    int length = snprintf(fullpath, sizeof(fullpath), "%s/%s", path, program->program_name);
+    if (length < 0) {
+        return -EINVAL;
+    } else if (length > sizeof(fullpath)) {
+        return -ENAMETOOLONG;
+    }
+
+    uint32_t result = ebpf_api_unpin_object((const uint8_t*)fullpath, (uint32_t)strlen(fullpath));
+    return (int)result;
+}
+
+static void
+_unpin_previous_programs(_In_ ebpf_program_t* program, _In_ struct bpf_object* object, const char* path)
+{
+    for (program = bpf_program__prev(program, object); program; program = bpf_program__prev(program, object)) {
+        bpf_program__unpin(program, path);
+    }
+}
+
+int
+bpf_program__pin(struct bpf_program* program, const char* path)
+{
+    char fullpath[MAX_PATH];
+    int result;
+
+    int length = snprintf(fullpath, sizeof(fullpath), "%s/%s", path, program->program_name);
```

Comment:
bpf_program__pin() expects full pin path, so we should not append program name to the path.

Reply:
Fixed, thanks